lib/improver/psychrometric_calculations/psychrometric_calculations.py

```diff
@@ -589,3 +594,144 @@ def process(self, temperature, relative_humidity, pressure):
+
+
+class FallingSnowLevel(object):
+    """Calculate a field of continuous falling snow level."""
+
+    def __init__(self, precision=0.005, falling_level_threshold=90.0):
+        """
+        Initialise class.
+
+        Keyword Args:
+            precision (float):
+                The precision to which the Newton iterator must converge
+                before returning wet bulb temperatures.
+            falling_level_threshold (float):
+                The cutoff threshold for the Wet-bulb integral used
+                to calculate the falling snow level.
+
+        """
+        self.precision = precision
+        self.wet_bulb_integral_plugin = (
+            WetBulbTemperatureIntegral(precision=precision))
+        self.falling_level_threshold = falling_level_threshold
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<FallingSnowLevel: precision:'
+                  '{}, falling_level_threshold:{}>'.format(
+                      self.precision,
+                      self.falling_level_threshold))
+        return result
+
+    def find_falling_level(self, wb_int_data, orog_data, height_points):
+        """
+        Find the falling snow level by finding the level of the wet-bulb
+        integral data at the required threshold.
+
+        Args:
+            wb_int_data (np.array):
+                Wet bulb integral data on heights
+            orog_data (np.array):
+                Orographic data
+            heights (np.array):
+                heights agl
+
+        Returns:
+            snow_level_data (np.array):
+                Falling snow level data asl.
+
+        """
+        # Create cube of heights above sea level for each height in
+        # the wet bulb integral cube.
+        asl = wb_int_data.copy()
+        for i, height in enumerate(height_points):
+            asl[i, ::] = orog_data + height
+
+        # Calculate falling snow level above sea level using
+        # monty.vinterp.interpolate.
+        # Interpolate returns a cube with height coord
+        #  for falling_level_threshold so we take the 0 index
+        snow_level_data = interpolate(np.array([self.falling_level_threshold]),
+                                      wb_int_data, asl, axis=0)[0]
+        return snow_level_data
+
+    @staticmethod
+    def fill_in_missing_data(snow_level_data):
+        """
+        Fill in missing data
+
+        Args:
+            snow_level_data (np.array):
+                Falling snow level data.
+
+        Returns:
+            snow_level_updated (np.array):
+                Falling snow level data with missing data added.
+
+        """
+        points = np.where(np.isfinite(snow_level_data))
+        values = snow_level_data[points]
+        ynum, xnum = snow_level_data.shape
+        (y_points, x_points) = np.mgrid[0:ynum, 0:xnum]
+        snow_level_updated = griddata(points, values, (y_points, x_points),
```

Comment:
Could we consider checking any remaining nan values after trying the "linear" approach, and for those points using the "nearest" method? I imagine the only reason we will get unfilled data is because we've hit ground, but setting it to zero says it would still be snow upon reaching sea level (I think), which is quite possibly not true and could lead to odd jumps in the snow fall level at the fringes of mountains etc.

Thoughts?

Reply:
I did try filling in the missing data with nearest after running the linear approach but it just made things worse as it picked up some random point and spread it over a large area. It is only at the edges when there is no data at the edge (i.e the snow is reaching the surface) that the linear method breaks down. This is seen in the top right hand corner over Norway.

Reply:
Having thought about this I think we want to replace any nan values after the linear method with the orography height at that point. In that way we don't suggest that the snow would reach sea-level at these points, but that it will reach ground level. Does that sound reasonable?